### DISCUSSION TITLE
How about Client.base_url and Request.url merge?

### DISCUSSION THREAD

**Discussion by Atralupus** [encode/httpx#1779] [Q&A] [answered]
I often use below code when using the requests library

```
from requests import Session
from urllib.parse import urljoin

class BaseUrlSession(Session):
    def __init__(self, base_url=None, *args, **kwargs):
        super(BaseUrlSession, self).__init__(*args, **kwargs)
        self.prefix_url = base_url

    def request(self, method, url, *args, **kwargs):
        url = urljoin(self.prefix_url, url)
        return super(BaseUrlSession, self).request(
            method, url, *args, **kwargs
        )
...

session = BaseUrlSession("https://foo.co.kr/")
session.get("/foo")

```

I like httpx Client's base_url properties
But httpx cloudn't use like above

```
import httpx

request = httpx.Request("GET", "/common.do")
client = httpx.Client(base_url="https://www.foo.co.kr/")
print(client.send(request).json()) # booom
```

If there is a base_url among the properties of the client when sending, How about there is a function to merge it with the url of the request?

**Comment by lovelydinosaur** [comment, marked as answer]
So, this'll work...

```python
client = httpx.Client(base_url="https://www.foo.co.kr/")
response = client.get("/common.do")
```

This would also work...

```python
client = httpx.Client(base_url="https://www.foo.co.kr/")
request = client.build_request("GET", "/common.do")
response = client.send(request)
```

The difference is that if you instantiate the request directly and then call `.send()`, then that's exactly what the client attempts to send, without making any further alterations to it.

Worth considering if our docs at https://www.python-httpx.org/advanced/#request-instances could do with any tweaking to help make things clearer here.

**Reply by Atralupus** [reply]
Thanks for the detailed explanation🥇 I was stupid

**Reply by lovelydinosaur** [reply]
You're most welcome. 😀
